### PR TITLE
OpenZFS 9075 - Improve ZFS pool import/load..

### PR DESCRIPTION
**Pull request by dweeezil**
### Description
Port the rest of the SPA loading-related changes from OpenZFS to ZoL.

### Motivation and Context
OpenZFS port.

### How Has This Been Tested?
Manually ran some of the new ZTS scripts.
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by rugubara** [comment]
Hello, @dweeezil 

I have a pool failing to import. (see [zfs-discuss thread](http://list.zfsonlinux.org/pipermail/zfs-discuss/2018-April/030921.html) )
Would it make sense trying this code on it? How safe is it?

Thanks for help. I'm really desperate trying to save these data.

**Comment by draghuram** [comment]
Hi, As I posted at
    http://list.zfsonlinux.org/pipermail/zfs-discuss/2018-April/030934.html
I am trying to debug a problem with pool import and someone suggested I try this patch. I am planning to do that but just wondering if anyone here can tell me what kind of improvements in error messages I can expect. Or if there is a debug procedure I should follow to get maximum benefit from the changes, it would be helpful to know that as well. Thanks.

**Comment by dweeezil** [comment]
@rugubara, @draghuram and other wanting to test this patch stack:  The focus of this work was to streamline the code paths involved with pool import, to provide better error messages during import, to better support various types of "rewind" imports and also to lay the groundwork for future enhancements.  If your troublesome pools only have a single top-level vdev, which appears to be the case in at least one of your reports, I doubt this patch stack would make any difference.

I'll also note that the patch stack appears to be stable, however, there are still issues with the test suite which need to be addressed before it will start passing the tests.  There were a bunch of tests added to the "zpool_import" area which unfortunately has a lot of unrelated baggage related to partitioning, etc.  It's currently passing all but the "import_rewind_config_changed"  test which we may simply have to disable because it is expected to fail in some cases.  I should have another push soon which ought to start passing most of the tests on the buildbots.

**Comment by dweeezil** [comment]
@behlendorf This looks much better.  Do I gather that the `zfs diff` problem is currently to be expected in the new Debian 9 tester?  Also, the problem in the Coverage tester is apparently due to a single new test using `sudo zpool` rather than simply `zpool`.

File: module/zfs/spa.c [hunk 1]
```diff
@@ -7563,9 +7863,18 @@ module_param(spa_load_verify_data, int, 0644);
...
+
 /* CSTYLED */
 module_param(zio_taskq_batch_pct, uint, 0444);
 MODULE_PARM_DESC(zio_taskq_batch_pct,
 	"Percentage of CPUs to run an IO worker thread");
 
+/* CSTYLED */
+module_param(zfs_max_missing_tvds, ulong, 0644);
+MODULE_PARM_DESC(zfs_max_missing_tvds,
+	"Maximum no. of missing top-level vdevs");
```
Comment (pzakha) [inline module/zfs/spa.c]: "Allow importing pool with up to this number of missing top-level vdevs (in read-only mode)."

File: module/zfs/spa.c [hunk 2]
```diff
@@ -7563,9 +7863,18 @@ module_param(spa_load_verify_data, int, 0644);
 MODULE_PARM_DESC(spa_load_verify_data,
 	"Set to traverse data on pool import");
 
+module_param(spa_load_print_vdev_tree, int, 0644);
+MODULE_PARM_DESC(spa_load_print_vdev_tree,
+	"Print vdev tree during pool import");
```
Comment (pzakha) [inline module/zfs/spa.c]: "Print vdev tree to zfs_dbgmsg during pool import"

File: tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import.kshlib [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+#
+# Use mdb to find the last txg that was synced in an active pool.
+#
+function get_last_txg_synced
+{
+	typeset pool=$1
+
+	if is_linux; then
+		txg=$(tail "/proc/spl/kstat/zfs/$pool/txgs" | awk '$3=="C" {print $1}' | tail -1)
```
Comment (pzakha) [inline tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import.kshlib]: nit: 80 characters per line

**Comment by dweeezil** [comment]
@pzakha Thanks for the review.  I'll get this all taken care of and also remove the `sudo` from the ZTS and get this re-pushed this afternoon.

**Pull request closed without merging**